### PR TITLE
Implement Connection Network Timeout

### PR DESCRIPTION
**Pull request by injectives**
The timeout value applies to individual server responses, not the JDBC API calls. Any response from the server, including the `NOOP` chunk, interrupts the idle period.

When initializing a new Bolt connection, the Neo4j server may supply a [connection.recv_timeout_seconds](https://neo4j.com/docs/bolt/current/appendix/connection-hints/#hint-recv-timeout-seconds) connection hint that defines the amount of time for which the connection may remain idle. When the network timeout value is set to `0` and the hint value is available, the latter is used as a default.

This update also fixes an issue when database name would not be supplied in the `BEGIN` message for explicit transaction. Also, the commit handler has been updated to handle a missing bookmark.

### CODE REVIEW

**Comment by injectives** [comment]
fixes #233

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ExtendedNeo4jConnection.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.neo4j.jdbc;
+
+/**
+ * Non public extension of {@link Neo4jConnection} used for internal purposes.
+ *
+ * @author Neo4j Drivers Team
+ * @since 6.0.0
+ */
+non-sealed interface ExtendedNeo4jConnection extends Neo4jConnection {
```
Comment (michael-simons) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/ExtendedNeo4jConnection.java]: Can we do without that interface, pls?

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/Neo4jConnection.java [hunk 2]
```diff
@@ -27,7 +29,52 @@
...
+	 * {@literal 0}, the connection hint value is used providing it was supplied by the
+	 * server or the timeout is turned off.
+	 * @throws SQLException if a database access error occurs, this method is called on a
+	 * closed connection or the value specified for seconds is less than 0.
+	 * @see <a href=
+	 * "https://neo4j.com/docs/bolt/current/appendix/connection-hints/#hint-recv-timeout-seconds">connection.recv_timeout_seconds
+	 * connection hint</a>
+	 */
+	@Override
+	void setNetworkTimeout(Executor executor, int milliseconds) throws SQLException;
```
Comment (michael-simons) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/Neo4jConnection.java:61]: Excellent work with explaining this.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/StatementImpl.java [hunk 3]
```diff
@@ -157,6 +157,7 @@ public void close() throws SQLException {
 		}
 		closeResultSet();
 		this.closed = true;
+		this.connection.onClose(this);
```
Comment (michael-simons) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/StatementImpl.java]: As the connection hands out the statement, it should keep track of it. If this is not possible, define a callback independent of the connection and don't intertwine both concepts. 

I think adding an `addCloseListenener` on the statementimpl class is more appropriate.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java [hunk 4]
```diff
@@ -66,12 +71,16 @@
...
-final class ConnectionImpl implements Neo4jConnection {
+final class ConnectionImpl implements ExtendedNeo4jConnection {
+
+	private static final Logger LOGGER = Logger.getLogger(Neo4jConnection.class.getCanonicalName());
 
 	private static final int TRANSLATION_CACHE_SIZE = 128;
 
 	private final BoltConnection boltConnection;
 
+	private final Set<AutoCloseable> instantiatedAutoClosables = new HashSet<>();
```
Comment (michael-simons) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java]: Naming: I think `trackedStatements` is better.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java [hunk 5]
```diff
@@ -661,6 +691,42 @@ public void flushTranslationCache() {
...
+	}
+
+	private void handleFatalException(SQLException fatalSqlException, SQLException sqlException) {
+		var cause = sqlException.getCause();
+		if (cause instanceof ConnectionReadTimeoutException) {
+			var iterator = this.instantiatedAutoClosables.iterator();
+			while (iterator.hasNext()) {
+				var autoClosable = iterator.next();
+				iterator.remove();
+				try {
```
Comment (michael-simons) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java]: Can pull the next into to the `try()`and do remove in the finally block. Feels better.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java [hunk 6]
```diff
@@ -661,6 +691,42 @@ public void flushTranslationCache() {
 		}
 	}
 
+	private <T extends AutoCloseable> T addAutoCloseable(T autoCloseable) {
```
Comment (michael-simons) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java]: Naming: "regiterStatement" or similar. Plus I would hook up the this connection as listener on the statement here and only here. So you have the machinery decoupled.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/DefaultTransactionImpl.java [hunk 7]
```diff
@@ -172,13 +164,18 @@ private <T> T execute(CompletableFuture<T> future, int timeout) throws SQLExcept
 		}
 		catch (ExecutionException ex) {
 			var cause = ex.getCause();
+			if (cause == null) {
+				cause = ex;
+			}
+			var sqlException = new SQLException("An error occurred while handling request.", cause);
```
Comment (michael-simons) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/DefaultTransactionImpl.java]:
```suggestion
			var sqlException = new SQLException("An error occurred while handling request", cause);
```

No dots at the end of the message (in line with JDK and Spring messages)

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/DefaultTransactionImpl.java [hunk 8]
```diff
@@ -172,13 +164,18 @@ private <T> T execute(CompletableFuture<T> future, int timeout) throws SQLExcept
...
 			if (cause instanceof Neo4jException || cause instanceof MessageIgnoredException) {
 				fail(new SQLException("The transaction is no longer valid."));
 			}
 			else {
-				fail(new SQLException("The connection is no longer valid."), true);
+				fail(new SQLException("The connection is no longer valid."));
```
Comment (michael-simons) [inline neo4j-jdbc/src/main/java/org/neo4j/jdbc/DefaultTransactionImpl.java]:
```suggestion
				fail(new SQLException("The connection is no longer valid"));
```

**Review by michael-simons — CHANGES_REQUESTED** [review]

**Comment by michael-simons** [comment]
Also before I forget over night: what about race conditions, connection closing the statement the statement trying to remove itself? Do we have tests for that?

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/Neo4jConnection.java (hunk 2 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/Neo4jConnection.java:61]: Thanks! 👍

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java (hunk 4 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java]: Good point, I was actually going to find a better name, but my focus moved to the timeout implementation.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java (hunk 5 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java:709]: Done. 👍

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/DefaultTransactionImpl.java (hunk 7 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/DefaultTransactionImpl.java]: Updated for this class.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/DefaultTransactionImpl.java (hunk 8 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/DefaultTransactionImpl.java]: Updated for this class.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java (hunk 6 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java]: Updated as agreed.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java (hunk 5 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java]: Updated as agreed.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/StatementImpl.java (hunk 3 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/StatementImpl.java]: Updated as agreed.

File: neo4j-jdbc/src/main/java/org/neo4j/jdbc/ExtendedNeo4jConnection.java (hunk 1 above)
Comment (injectives) [reply neo4j-jdbc/src/main/java/org/neo4j/jdbc/ExtendedNeo4jConnection.java]: Deleted. 👍

**Comment by injectives** [comment]
> Also before I forget over night: what about race conditions, connection closing the statement the statement trying to remove itself? Do we have tests for that?

Since we've agreed to update the tracking logic, this should not be an issue.

**Review by michael-simons — APPROVED** [review]
Great fun coming to this solution together, thank you!

**Pull request merged**